### ISSUE TITLE
Stubbing some DOM methods can halt the test suite or cause a fatal error

### ISSUE THREAD

**Issue by andyearnshaw** [theintern/intern#390]
I've noticed that some DOM methods can cause problems for the runner, even if they're restored in teardown/afterEach.  An example:

```javascript
var xhr;
registerSuite({
beforeEach: function () {
xhr = sinon.useFakeXMLHttpRequest();
},
afterEach: function () {
xhr.restore()
},
'test that some method tries to send a request': function () {
// ...
}
});
``` Intern hangs for me after this test runs, presumably because it's trying to use the real `XMLHttpRequest` constructor to send a real request and expects a certain response.  I've seen that stubbing `HTMLElement.prototype.appendChild` can cause similar problems, but perhaps that's less likely to crop up.  Still it might be worth fixing at the same time, as it's straightforward enough to just keep an internal reference to these methods before the tests are run.

**Comment by csnover** [comment]
Well, this is… awkward.

The runner sends data back to the client using `dojo/request`, which is an abstraction. It doesn’t use an XMLHttpRequest object directly, so we can’t just retain a reference. I guess we’ll need to update `dojo/request` upstream to accept being passed an XMLHttpRequest constructor, instead of getting it from the global scope, and then have the reporter retain a copy of that constructor and pass it explicitly for every request. It’s doable, but I’m not sure what the timeline would be to actually fix it without a support plan.

As far as the mutation of DOM prototypes goes… what are you testing that does such a thing? I’m happy to implement more defensiveness in some of the code but I don‘t want to introduce complexity for theoretical problems. (Also the DOM prototypes thing, until pretty recently, wasn’t a thing you could even really do, since old IE and Chrome didn’t conform to WebIDL.)

**Comment by andyearnshaw** [comment]
Yeah, after some more investigation I realised just before I left for home
that the issue was in `dojo/request`, which does complicate things
somewhat.  I was able to work around the issue in the meantime by using
sinon's clever built-in filters and applying the fake requests only to
request URLs that started with a term that would likely be unique to my
tests.

WRT DOM prototypes, Internet Explorer 8 exposed `Element` and IE 9 exposed
`HTMLElement`, and AFAIK Chrome has exposed them since version 1.  My
reasons for using it were to test a function that adds scripts to the
document, checking that the script is added with the correct src and other
attributes.  I also attempted to stub the method directly on the `<head>`
element, but I've hit what looks like a bug in Phantom (works fine in
Chrome) where the stub doesn't get restored properly and it affects another
test later down the line.
